mono/utils/mono-safe-handle.c

```diff
@@ -0,0 +1,167 @@
+/**
+ * \file
+ * Support for synchronously freeing unmanaged memory subject to
+ * concurrent refcount modification
+ *
+ * Author:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * (C) 2018 Microsoft, Inc.
+ *
+ */
+
+#include <config.h>
+#include <glib.h>
+#include "utils/mono-safe-handle.h"
+#include "utils/atomic.h"
+
+#define INTERNAL_HANDLE_TOMBSTONE (GINT_TO_POINTER (-1))
+
+static void 
+ensure_synch_cs_set (MonoSafeHandleNative *handle)
+{
+	MonoCoopMutex *synch_cs;
+
+	if (handle->synch_cs != NULL) {
+		return;
+	}
+
+	synch_cs = g_new0 (MonoCoopMutex, 1);
+	mono_coop_mutex_init_recursive (synch_cs);
```

Comment:
Need barrier after `mono_coop_mutex_init_recursive`, so its affects are done before the atomic_cas_ptr? Or is one or the other already a full barrier?

In either case..I have to ask in slack...there is a theory that there are two missing barriers here, but maybe not.

Reply:
I thought about it -- I suggest a barrier between the init and the cas.
So the init will have finished before the cas, so another thread doesn't see a pointer that isn't initialized.

Reply:
Er, and before the return, to avoid prefetching from lock. Let me double  check.